### PR TITLE
[node-bridge] Add `res.waitUntil` for lambdas and wait for handler to finish

### PR DESCRIPTION
**Pull request by karaggeorge**
An attempt to add `waitUntil` to node lambdas and to prevent code execution from getting cut after `res.end()`

**DISCLAIMER**: this is my first time reading into the node/node-bridge packages, and I haven't found a good way to run/test this locally, so I might have missed some edge cases

From reading into node-bridge, I realized the reason why the execution of the handler can get cut-off is because in the bridge, we only wait for the response to be streamed back to end the event, but not for the actual handler to finish.

I tried to add a `waitUntil` helper that gets attached to the response backed by the node-bridge for a specific request by id.
It also adds the handler itself, so it can be awaited before finishing execution.

There's a few areas that I'm unsure about:

### 1

```
const req = request(
        { hostname: '127.0.0.1', port, path, method },
        socket && url && cipher
          ? getStreamResponseCallback({ url, socket, cipher, resolve, reject })
          : getResponseCallback({ isApiGateway, resolve, reject })
      );
```

Here, I'm not sure when each case happens, but reading through it, it seems like if `getStreamResponseCallback` is used, this will work as expected and stream the response back, while waiting to finish the tasks afterwards. But if `getResponseCallback` is used, then it'll still wait for the tasks to finish but it'll also hold the response until those are done.

### 2

In `node-bridge/launcher.js`, in `getListener` there's a 3rd case for exporting an entire Server object, or even having a server listen automatically. In those cases this wrapper won't apply, since it wraps the listener.

I _could_ try to make it work by wrapping the prototype of `ServerResponse` like we do for the `Server`, but I wasn't sure if this is something legacy that we don't need to worry about before I go down that route.

### 3

I _think_ this will automatically work for Next.js as well, since from what I see the lambda just uses this launcher as well, but it'll probably need updated types and actual implementation of a similar functionality in the next.js repo as well, to handle the dev/prod server.

#### Tests

- [ ] The code changed/added as part of this PR has been covered with tests
- [ ] All tests pass locally with `yarn test-unit`

#### Code Review

- [ ] This PR has a concise title and thorough description useful to a reviewer
- [ ] Issue from task tracker has a link to this PR

### CODE REVIEW

File: packages/node-bridge/launcher.js [hunk 1]
```diff
@@ -87,19 +87,29 @@ function getVercelLauncher({
...
           Server.prototype.listen = originalListen;
+
           if (shouldAddHelpers) {
             bridge.setStoreEvents(true);
-            import(helpersPath).then(helper => {
-              const h = helper.default || helper;
-              const server = h.createServerWithHelpers(listener, bridge);
+          }
+
+          import(helpersPath).then(helper => {
```
Comment (karaggeorge) [inline packages/node-bridge/launcher.js:95]: I added the new wrapper in `helpers.ts` mostly to add type-safety, but this means we have to import this in both cases of `shouldAddHelpers`. I could move that function in this file to avoid having to do that

File: packages/node-bridge/bridge.js [hunk 2]
```diff
@@ -205,6 +212,25 @@ class Bridge {
     );
   }
 
+  /**
+   * @param {string} waitUntilId
+   * @param {Promise<any>} promise
+   * @param {boolean} isHandler
+   */
+  waitUntil(waitUntilId, promise, isHandler) {
```
Comment (ijjk) [inline packages/node-bridge/bridge.js:220]: I wonder if this should be allowed in non-streaming mode as it may cause confusion why the response is delayed 🤔

**Review by ijjk — COMMENTED** [review]
Is there a new thread related to this? The last one that we discussed this in was here https://vercel.slack.com/archives/C0457JK07KP/p1665720487686199 not sure if things have change since then.

**Comment by karaggeorge** [comment]
@ijjk had no idea about that thread/project, was just something I've been needing more recently, and decided to try implementing for DYF

File: packages/node-bridge/bridge.js (hunk 2 above)
Comment (karaggeorge) [reply packages/node-bridge/bridge.js:220]: Yeah, I noticed that too. What are the parameters that determine which scenario happens? Streaming vs not? It wasn't clear in the code

Comment (ijjk) [reply packages/node-bridge/bridge.js:220]: It's only streaming when `responseCallbackUrl` is present currently.

**Comment by EndangeredMassa** [comment]
Revamped in: https://github.com/vercel/vercel/pull/9745
